components/Typography/src/UIFont+MaterialScalable.m

```diff
@@ -0,0 +1,107 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "UIFont+MaterialScalable.h"
+
+#import <objc/runtime.h>
+
+#import "MDCTypography.h"
+#import "UIApplication+AppExtensions.h"
+
+static char MDCFontScaleObjectKey;
+
+@implementation UIFont (MaterialTypography)
+
+- (UIFont *)mdc_scaledFontForSizeCategory:(UIContentSizeCategory)sizeCategory {
+  if (!self.mdc_scalingCurve) {
+    return self;
+  }
+
+  NSNumber *fontSizeNumber;
+  if (sizeCategory) {
+    fontSizeNumber = self.mdc_scalingCurve[sizeCategory];
+  }
+
+  // Guard against broken / incomplete scaling curves by returning self if fontSizeNumber is nil.
+  if (fontSizeNumber == nil) {
+    return self;
+  }
+
+  CGFloat fontSize = (CGFloat)fontSizeNumber.doubleValue;
+  UIFont *scaledFont = [UIFont fontWithDescriptor:self.fontDescriptor size:fontSize];
+  scaledFont.mdc_scalingCurve = self.mdc_scalingCurve;
+
+  return scaledFont;
+}
+
+/**
+ @return Device's current UIContentSizeCategory or UIContentSizeCategoryLarge
+     if we are unable to query the device due to being in an extension.
+ */
+static UIContentSizeCategory GetCurrentSizeCategory(void) {
+  UIContentSizeCategory sizeCategory = UIContentSizeCategoryLarge;
+  if ([UIApplication mdc_safeSharedApplication]) {
+    sizeCategory = [UIApplication mdc_safeSharedApplication].preferredContentSizeCategory;
+  } else if (@available(iOS 10.0, *)) {
+    sizeCategory = UIScreen.mainScreen.traitCollection.preferredContentSizeCategory;
+  }
+
+  return sizeCategory;
+}
+
+- (UIFont *)mdc_scaledFontForCurrentSizeCategory {
+  UIContentSizeCategory currentSizeCategory = GetCurrentSizeCategory();
+
+  return [self mdc_scaledFontForSizeCategory:currentSizeCategory];
+}
+
+- (nonnull UIFont *)mdc_scaledFontAtDefaultSize {
+  return [self mdc_scaledFontForSizeCategory:UIContentSizeCategoryLarge];
+}
+
+- (CGFloat)mdc_scaledValueForValue:(CGFloat)value {
+  UIContentSizeCategory defaultSizeCategory = UIContentSizeCategoryLarge;
+  UIContentSizeCategory currentSizeCategory = GetCurrentSizeCategory();
+
+  NSNumber *defaultFontSizeNumber = self.mdc_scalingCurve[defaultSizeCategory];
+
+  NSNumber *currentFontSizeNumber = self.mdc_scalingCurve[currentSizeCategory];
+  // Guard against broken / incomplete scaling curves by returning self if fontSizeNumber is nil.
+  if (currentFontSizeNumber == nil || defaultFontSizeNumber == nil) {
+    return value;
+  }
+
+  CGFloat currentFontSize = (CGFloat)currentFontSizeNumber.doubleValue;
+  CGFloat defaultFontSize = (CGFloat)currentFontSizeNumber.doubleValue;
+
+  // Guard against broken / incomplete scaling curves by returning original value if the
+  // fontSize <= 0.0.
+  if (currentFontSize <= 0.0 || defaultFontSize <= 0.0) {
+    return value;
+  }
+
+  return currentFontSize / defaultFontSize;
```

Comment:
This looks like duplicated code from MDCFontScaler. I'm also surprised it doesn't use `value` in the expected return value.

Reply:
Fixed.